llvm/lib/Support/SimpleTable.cpp

```diff
@@ -0,0 +1,247 @@
+    if (Names.erase(StringRef(*Iter)) > 0)
+      continue; // yes
+    // no - remove from titles (ColumnNum2Name will be updated in rebuild below)
+    ColumnName2Num.erase(*Iter);
+    ColumnNames.erase(Iter);
+    // ... and from data
+    for (int Row = 0; Row < getNumRows(); ++Row)
+      Rows[Row].Cells.erase(Rows[Row].Cells.begin() + Col);
+  }
+  if (Names.size() > 0)
+    return makeError("column not found " + *Names.begin());
+  rebuildName2NumMapping();
+  return Error::success();
+}
+
+void SimpleTable::linearize(std::vector<std::string> &Res) const {
+  for (const auto &R : Rows)
+    for (const auto &C : R.Cells)
+      Res.push_back(C);
+}
+
+static constexpr char COL_TITLE_LINE_OPEN[] = "[";
+static constexpr char COL_TITLE_LINE_CLOSE[] = "]";
+static constexpr char ROW_SEP[] = "\n";
+
+void SimpleTable::write(raw_ostream &Out, bool WriteTitles, char ColSep) const {
+  if (WriteTitles) {
+    Out << COL_TITLE_LINE_OPEN;
+
+    for (unsigned I = 0; I < ColumnNames.size(); ++I) {
+      if (I != 0)
+        Out << ColSep;
+      Out << *ColumnNum2Name[I];
+    }
+    Out << COL_TITLE_LINE_CLOSE << ROW_SEP;
+  }
+  const unsigned N = ColumnNames.size();
+
+  for (unsigned I = 0; I < Rows.size(); ++I) {
+    const auto &R = Rows[I];
+
+    for (unsigned J = 0; J < N; ++J) {
+      if (J != 0)
+        Out << ColSep;
+      Out << R.Cells[J];
+    }
+    Out << ROW_SEP;
+  }
+}
+
+Expected<SimpleTable::UPtrTy> SimpleTable::read(MemoryBuffer *Buf,
+                                                char ColSep) {
+  int Cnt = 0;
+  line_iterator LI(*Buf);
+
+  if (LI.is_at_end() || LI->empty()) // empty table
+    return std::make_unique<SimpleTable>();
+  ++Cnt;
+  std::unique_ptr<SimpleTable> Res;
+
+  if (LI->startswith(COL_TITLE_LINE_OPEN)) {
+    if (!LI->endswith(COL_TITLE_LINE_CLOSE))
+      return createStringError(errc::invalid_argument, "malformed title line");
+    // column titles present
+    auto L = LI->substr(1, LI->size() - 2); // trim '[' and ']'
+    SmallVector<StringRef, 4> Titles;
+    L.split(Titles, ColSep);
+    auto Table = SimpleTable::create(Titles);
+    if (!Table)
+      return Table.takeError();
+    Res = std::move(Table.get());
+    LI++;
+    ++Cnt;
+  }
+  // parse rows
+  while (!LI.is_at_end()) {
+    SmallVector<StringRef, 4> Vals;
+    LI->split(Vals, ColSep);
+
+    if (!Res) {
+      auto Table = SimpleTable::create(Vals.size());
+      if (!Table)
+        return Table.takeError();
+      Res = std::move(Table.get());
+    }
+    if (static_cast<int>(Vals.size()) != Res->getNumColumns())
+      return createStringError(errc::invalid_argument,
+                               "row size mismatch at line " + Twine(Cnt));
+    Res->addRow(Vals);
+    LI++;
+    ++Cnt;
+  }
+  return std::move(Res);
+}
+
+Expected<SimpleTable::UPtrTy> SimpleTable::read(const Twine &FileName,
+                                                char ColSep) {
+  auto MemBuf = MemoryBuffer::getFileAsStream(FileName);
```

Comment:
First, It is not obvious what `getFileAsStream` returns.
LLVM coding guide suggests use `auto` only if type is obvious.
```suggestion
  llvm::ErrorOr<std::unique_ptr<llvm::MemoryBuffer>> MemBuf = MemoryBuffer::getFileAsStream(FileName);
```
Second
I suggest leave "reading from file" functionality implementation for corresponding tool.
Just to be logical, why the data table would know how to read files from disk?
Second, LLVM provides better error handing for tools, see http://llvm.org/docs/ProgrammersManual.html#interoperability-with-std-error-code-and-erroror .

Reply:
Re second - it is really a convenience for the users. Every tool will then implement it. I think this functionality should remain.